keps/sig-auth/20190408-node-scoped-daemonset.md

```diff
@@ -0,0 +1,239 @@
+    * Pros: this solution will work properly for the cases where a single
+      service account is used to run multiple DaemonSets which have different
+      scope requirements.
+    * Cons: this solution is complex to implement. A new boolean field in
+      [TokenRequestSpec](https://github.com/kubernetes/kubernetes/blob/dd96b15e5d1de1923645ef17c05dfdfb616163bf/staging/src/k8s.io/api/authentication/v1/types.go#L142)
+      and a new boolean field in
+      [serviceaccount.privateClaims.Kubernetes](https://github.com/kubernetes/kubernetes/blob/46c8b075ded386d99de1cf3945168127403c12e9/pkg/serviceaccount/claims.go#L38)
+      will be required to integrate the node-scoped info into service account
+      tokens.
+
+Considering that it is a best practice to run a DaemonSet using a dedicated
+service account, the first option is chosen.
+
+### Workflow of NodeAuthorizer for DaemonSet Requests
+
+* A DaemonSet pod from Node N1 sends a request to APIServer;
+* APIServer authenticates the request,
+  [adds](https://github.com/kubernetes/kubernetes/blob/255a73ffc64d1c4e0f9a378ebf8d0005edfe6962/pkg/serviceaccount/util.go#L55)
+  the pod name, pod UID, and node-scoped info into the
+  [Extra](https://github.com/kubernetes/kubernetes/blob/255a73ffc64d1c4e0f9a378ebf8d0005edfe6962/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L50)
+  field of
+  [user.DefaultInfo](https://github.com/kubernetes/kubernetes/blob/255a73ffc64d1c4e0f9a378ebf8d0005edfe6962/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L46)
+  (require changing
+  [pkg/serviceaccount](https://github.com/kubernetes/kubernetes/tree/master/pkg/serviceaccount));
+* NodeAuthorizer gets the pod info and the node-scoped info from the Extra field
+  of user.DefaultInfo through
+  [GetExtra](https://github.com/kubernetes/kubernetes/blob/d99f49d2a7caa2e86cb27541a267577771e133a7/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L41);
+  using the pod info, NodeAuthorizer can get the node information of the request
+  sender by checking its resource graph (require changing
+  [pkg/auth/nodeidentifier](https://github.com/kubernetes/kubernetes/tree/master/pkg/auth/nodeidentifier)
+  and NodeAuthorizer);
+* At this point, NodeAuthorizer has all the necessary information to make its
+  authorization decision. The graph below shows the decision making process.
+
+![The decision-making process of NodeAuthorizer](node-scoped-daemonset.png)
+
+### Workflow of NodeRestriction for DaemonSet Requests
+
+* A DaemonSet pod P1, from Node N1 sends a request to APIServer;
+* APIServer authenticates the request, adds the pod name, the pod UID, and the
+  node-scoped info into the Extra field of user.DefaultInfo;
+* NodeRestriction gets the pod info and the node-scoped info from the Extra
+  field of user.DefaultInfo through GetExtra; using the pod info,
+  NodeRestriction can gets the node information of P1 by checking
```

Comment:
I thought it could just fetch it directly from the graph?

Reply:
Do you mean the resource graph used in NodeAuthorizer?

Since NodeRestriction can inspect the request objects, there is no need to check the resource graph.